File: pipeline/InputHandler.m
classdef InputHandler < handle
    % InputHandler: Handles input configuration and processing of MF4 files
    
    properties
        jsonConfigPath      % Path to JSON config file
        pathToRawData       % Path to raw data folder - *.mf4, set in constructor
        % Mirrored Config properties
        signalMap
        graphSpec           % Table from sheet 'graphSpec'
        lineColors          % Table from sheet 'lineColors'
        % Additional properties for convenience
        signalPlotSpecPath
        signalPlotSpecName
    end

    methods
        function obj = InputHandler(config)
            % Constructor: accepts a Config object and sets pathToRawData
            if nargin < 1
                error('Configuration object is required for initialization.');
            end
            
            % Validate input is a Config instance
            if ~isa(config, 'Config')
                error('Input argument must be an instance of Config.');
            end
            
            % Mirror Config properties
            obj.jsonConfigPath      = config.jsonConfigPath;
            obj.graphSpec           = config.graphSpec;
            obj.lineColors          = config.lineColors;
            obj.signalMap           = config.signalMap;
            obj.signalPlotSpecPath  = config.signalPlotSpecPath;
            obj.signalPlotSpecName  = config.signalPlotSpecName;

            % Prompt user to select folder with MF4 files
            curFolder = pwd;
            pathToRawData = uigetdir(curFolder, 'Select folder containing MF4 files');
            if pathToRawData == 0
                error('No folder selected. Operation cancelled.');
            end

            % Set pathToRawData to the selected folder
            obj.pathToRawData = pathToRawData;

            % Verify if folder contains MF4 files
            files = dir(fullfile(obj.pathToRawData, '*.mf4'));
            if isempty(files)
                error('No MF4 files found in selected folder: %s', obj.pathToRawData);
            end

            % Verify if folder contains MF4 files
            files = dir(fullfile(obj.pathToRawData, '*.mf4'));
            if isempty(files)
                error('No MF4 files found in folder: %s', obj.pathToRawData);
            end
        end

        function processedData = processMF4Files(obj)
            % Process MF4 files: extract specified signals, save to MAT, and return processed data
            curFolder = pwd;
            
            % Use pre-set pathToRawData
            pathToRawData = obj.pathToRawData;

            % Verify if folder contains MF4 files
            files = dir(fullfile(pathToRawData, '*.mf4'));
            if isempty(files)
                fprintf('No MF4 files found in selected folder: %s\n', pathToRawData);
                cd(curFolder);
                processedData = {};
                return;
            end

            % Initialize cell array to store processed data
            processedData = cell(1, length(files));
            N = length(files);
            fprintf('    Found %d MF4 file(s) to process...\n', N);

            % Process each MF4 file
            cd(pathToRawData);
            for i = 1:N
                fullPath = fullfile(files(i).folder, files(i).name);
                [~, name, ~] = fileparts(fullPath);
                
                try
                    % Call internal method to process MF4 file, passing signalMap as table
                    signalMat = obj.processMF4FileInternally(fullPath, obj.signalMap, 0.01); % Resample to 100 Hz

                    % Save to .mat file in the same folder as MF4 files
                    matFileName = fullfile(pathToRawData, [name '.mat']);
                    save(matFileName, 'signalMat');
                    % fprintf('Processed: %s, saved to %s\n', files(i).name, matFileName);

                    % Store processed data
                    processedData{i} = signalMat;
                catch err
                    fprintf('Error processing %s: %s\n', files(i).name, err.message);
                    processedData{i} = [];
                end
            end

            cd(curFolder);
        end
    end

    methods (Access = private)
        function signalMats = processMF4FileInternally(~, filePath, signalMap, resampleRate)
            % Process MF4 file with Docker + Python
            %
            % Inputs:
            %   filePath     - Full path to the .mf4 file
            %   signalMap    - Table containing signal mapping data
            %   resampleRate - Resampling interval in seconds (e.g., 0.01 for 100 Hz)
            %
            % Outputs:
            %   signalMats - Data loaded from the generated .mat file

            if nargin < 3
                error('Usage: processMF4FileInternally(filePath, signalMap, resampleRate)');
            end
            if nargin < 4
                resampleRate = 0.01; % Default 10 ms
            end
            validateattributes(resampleRate, {'numeric'}, {'positive','scalar'}, ...
                'processMF4FileInternally', 'resampleRate');
            validateattributes(signalMap, {'table'}, {}, ...
                'processMF4FileInternally', 'signalMap');

            % Verify the .mf4 file exists
            if ~isfile(filePath)
                error('MF4 file not found: %s', filePath);
            end

            % Extract folder and base name
            [folder, baseName, ~] = fileparts(filePath);
            matFullPath = fullfile(folder, [baseName '.mat']);

            % Save signalMap table as a temporary CSV file
            tempCsvPath = fullfile(folder, 'tempSignalMap.csv');
            try
                writetable(signalMap, tempCsvPath);
            catch err
                error('Failed to write signalMap to temporary CSV: %s', err.message);
            end

            % Build Docker command
            dockerBin   = '/usr/local/bin/docker';
            projectRoot = fileparts(folder); % Go up from rawdata to project root

            dockerCmd = sprintf([ ...
                '%s run --rm -v "%s":/data mdf-python:latest ' ...
                'python /data/AEB_KPI_Project/docker/mdf2matConv.py "/data/rawdata/%s.mf4" ' ...
                '--signal_db "/data/rawdata/%s" --resample %g' ...
            ], dockerBin, projectRoot, baseName, 'tempSignalMap.csv', resampleRate);

            dockerCmd = strtrim(dockerCmd);

            % Run Docker
            status = system(dockerCmd);
            % Clean up temporary CSV file
            if isfile(tempCsvPath)
                delete(tempCsvPath);
            end
            if status ~= 0
                error('Docker failed to process %s', filePath);
            end

            % Load the generated .mat file
            if ~isfile(matFullPath)
                error('MAT file not generated: %s', matFullPath);
            end

            loaded = load(matFullPath);

            if isfield(loaded, 'signalMat')
                signalMats = loaded.signalMat;
            else
                error('MAT file did not contain expected variable "signalMat".');
            end
        end % processMF4FileInternally
    end
end

File: pipeline/input_handler.py
import os
import warnings
import numpy as np
from typing import List
from asammdf import MDF, Signal
from utils.mf4_extractor import mf4_extractor
from utils.signal_filters import accel_filter
from utils.enum_loader import EnumMapper
from utils.load_params import load_params_from_class, load_params_from_config

# For folder selection
import tkinter as tk
from tkinter import filedialog


class InputHandler:
    """
    InputHandler
    -------------
    Handles input configuration and processing of MF4 files.
    """

    PARAM_SPECS = {
        "resample_rate": {"default": 0.01, "type": float, "desc": "Resampling interval in seconds (e.g., 0.01 = 100 Hz)"},
        "cutoff_freq": {"default": 10.0, "type": float, "desc": "Low-pass filter cutoff frequency in Hz"},
    }

    def __init__(self, config):
        """
        Constructor: accepts a Config object, loads key parameters,
        and prompts the user to select the folder containing MF4 files.
        """

        # --- Validate config object ---
        if config is None:
            raise ValueError("Configuration object is required for initialization.")
        if not hasattr(config, "signal_map"):
            raise TypeError("Config object must define signal_map")

        # --- Assign base attributes ---
        self.signal_map         = config.signal_map
        self.path_to_raw_data   = None
        self.path_to_extracted  = None

        # --- Load parameters from class and then override with config ---
        load_params_from_class(self)
        if config:
            load_params_from_config(self, config)

        # --- Prompt user for MF4 folder ---
        print("📂 Please select the MF4 folder...")
        root = tk.Tk()
        root.withdraw()  # hide main window
        folder = filedialog.askdirectory(title="Select MF4 Folder")

        if not folder:
            raise ValueError("No MF4 folder selected. Aborting.")
        
        self.path_to_raw_data = os.path.abspath(folder)
        print(f"✅ Selected MF4 folder: {self.path_to_raw_data}")

        # --- Create subfolder for extracted files ---
        self.path_to_extracted = os.path.join(self.path_to_raw_data, "extracted")
        os.makedirs(self.path_to_extracted, exist_ok=True)

    # -------------------- Public API -------------------- #
    def process_mf4_files(self) -> None:
        """
        Process MF4 files:
        1. Extract specified signals using mf4_extractor()
        2. Filter longActAccel & latActAccel
        3. Convert egoSpeed (m/s → km/h)
        4. Save all signals to '_extracted.mf4'
        """

        mf4_files = [f for f in os.listdir(self.path_to_raw_data) if f.lower().endswith(".mf4")]
        print(f"🔎 Found {len(mf4_files)} MF4 file(s) to process...")

        if not mf4_files:
            print("⚠️ No MF4 files found in the selected folder.")
            return

        for file in mf4_files:
            full_path = os.path.join(self.path_to_raw_data, file)
            print(f"\n📂 Processing file: {file}")

            try:
                # --- 1️⃣ Extract signals ---
                data, mdf_obj, sigs, summary, used, raster, mods = mf4_extractor(
                    dat_path=full_path,
                    signal_database=self.signal_map,
                    req=None,
                    resample=self.resample_rate,
                    convert_to_tact_unit=True,
                )

                if data.empty:
                    print(f"⚠️ Warning: no data extracted from {file}")
                    continue

                # --- 2️⃣ Filter long & lat acceleration ---
                for sig in ["longActAccel", "latActAccel"]:
                    if sig in data.columns:
                        try:
                            sig_flt = accel_filter(
                                data.index.values,
                                data[sig].values,
                                cutoff_freq=self.cutoff_freq
                            )
                            data[f"{sig}Flt"] = sig_flt
                            print(f"   ✅ Filtered signal: {sig} → {sig}Flt")
                        except Exception as e:
                            warnings.warn(f"⚠️ Failed to filter {sig}: {e}")
                    else:
                        warnings.warn(f"⚠️ Signal '{sig}' not found in extracted data.")
                # --- 3️⃣ Convert egoSpeed from m/s to km/h ---
                if "egoSpeed" in data.columns:
                    try:
                        data["egoSpeedKph"] = data["egoSpeed"] * 3.6
                        print("   ✅ Converted egoSpeed → egoSpeedKph (m/s → km/h)")
                    except Exception as e:
                        warnings.warn(f"⚠️ Failed to convert egoSpeed: {e}")
                else:
                    warnings.warn("⚠️ Signal 'egoSpeed' not found in extracted data.")

                # --- 4️⃣ Save both raw + filtered signals to new MDF ---

                mapper = EnumMapper("config/enum_definitions.yaml")
                
                new_mdf = MDF()
                for col in data.columns:
                    try:
                        series = data[col]
                        if series.dtype == object:
                            enum_name = mapper.get_enum_for_signal(col)

                            if enum_name:
                                # Use YAML enum mapping
                                enum_table = mapper.enums.get(enum_name, {})
                                encoded = series.astype(str).map(lambda v: enum_table.get(v, np.nan))
                                if encoded.isna().any():
                                    missing = series[encoded.isna()].unique().tolist()
                                    print(f"⚠️ Unmapped values in {col}: {missing}")
                                samples = encoded.fillna(-1).astype(np.int16).to_numpy()

                                sig = Signal(
                                    samples=samples,
                                    timestamps=data.index.values,
                                    name=col,
                                    unit="u[1]",
                                    comment=f"Enum mapping: {enum_name}"
                                )
                            else:
                                # fallback categorical encoding
                                categories, encoded = np.unique(series.astype(str), return_inverse=True)
                                sig = Signal(
                                    samples=encoded.astype(np.int16),
                                    timestamps=data.index.values,
                                    name=col,
                                    unit="u[1]",
                                    comment=f"Categorical mapping: {dict(enumerate(categories))}"
                                )

                        else:
                            sig = Signal(
                                samples=series.to_numpy(dtype=np.float64),
                                timestamps=data.index.values,
                                name=col,
                                unit="u[1]"
                            )

                        new_mdf.append(sig)

                    except Exception as e:
                        print(f"⚠️ Failed to append signal {col}: {e}")

                # Extract just the base filename, e.g., "test.mf4" → "test_extracted.mf4"
                base_name = os.path.splitext(os.path.basename(full_path))[0] + "_extracted.mf4"

                # Build the new path inside the extracted folder
                extracted_file = os.path.join(self.path_to_extracted, base_name)

                # Save to the extracted folder
                new_mdf.save(extracted_file, overwrite=True)
                
                print(f"💾 Saved extracted + filtered signals → {extracted_file}")

            except Exception as e:
                print(f"❌ Error processing {file}: {e}")
